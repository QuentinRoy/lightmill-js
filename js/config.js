define({
    SERVER_ADDRESS: 'http://0.0.0.0:5000',
    TARGET_EXPERIMENT: 'nonmode'
});
