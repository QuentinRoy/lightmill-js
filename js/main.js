/*jslint browser:true*/
/*global requirejs, define */

requirejs.config({
    enforceDefine: true,
    paths: {
        'jquery': 'libs/jquery-2.0.0',
        'raphaeljs': 'libs/raphael-arrow-fix',
        'text': 'libs/text',
        'jstools': 'packages/jstools/src',
        'templates': '../templates',
        'state-machine': 'libs/state-machine-modified',
        'signals': 'libs/signals',
        'cookies': 'libs/cookies',
        'underscore': 'libs/underscore',
        'color': 'libs/color-0.4.4',
    },
    packages: [
        {
            name: 'sigmamenu',
            location: 'packages/sigmamenu/src/',
            main: 'sigma-menu'
        }
    ],
    shim: {
        underscore: {
            exports: '_'
        },

        color: {
            exports: 'Color'
        },
    }
});


define(['xpclient/manager', 'xpclient/tasks/task-factory'], function (XpManager, TaskFactory) {
    'use strict';
    var taskFactory = new TaskFactory(),
        manager = new XpManager(taskFactory);
    manager.start();
});