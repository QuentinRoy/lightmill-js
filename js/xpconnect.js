define(['config', 'jquery'], function (config, $) {

    function XpConnect() {
        this.serverAddress = config.SERVER_ADDRESS;
        this.targetXp = config.TARGET_EXPERIMENT;
        this.runId = null;

        this._request_run();
    }

    XpConnect.prototype = {

        _request_run: function () {
            var address = this.serverAddress + '/experiment/' + this.targetXp + '/next_run';
            $.get(address).done($.proxy(this._set_run, this)).fail(function () {
                console.log('Cannot get target run for experiment ' + this.targetXp + ' at ' + address);
            });
        },

        _set_run: function (runId) {
            this.runId = runId;
        }
    };

    return new XpConnect();

});