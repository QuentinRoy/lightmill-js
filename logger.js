define(['jstools/tools', 'jquery', 'fiber', 'jstools/dot-syntax', 'is'], function(tools, $, Fiber, dot, is) {
    'use strict';

    return Fiber.extend(function() {

        return {

            init: function(processor_s, settings) {
                this._log = {};
                settings = settings || {};

                if (is.fn(processor_s)) {
                    this.processors = [processor_s];
                } else if (is.array(processor_s)) {
                    this.processors = processor_s.slice(0);
                } else if (is.hash(processor_s)) {
                    this.processors = [];
                    for(var proc in processor_s) if (processor_s.hasOwnProperty(proc)) {
                        this.processors.push(processor_s[proc]);
                    }
                }

                // TODO: use this to protect listed values (mark them as unmodifiable)
                this.protectedValues = settings.protectedValues || [];
                this.processorParams = settings.processorParams;
            },


            timestamp: function(measureNameOrArrayOfMeasureNames) {
                var timeStamp = new Date().getTime(),
                    _this = this;
                if (typeof measureNameOrArrayOfMeasureNames === 'string') {
                    this.set(measureNameOrArrayOfMeasureNames, timeStamp);
                } else {
                    $.each(measureNameOrArrayOfMeasureNames, function(num, val) {
                        if (typeof val !== 'string')
                            throw 'measureName must be a string';
                        _this.set(val, timeStamp);
                    });
                }
            },

            get: function(measurePath) {
                return dot.get(this._log, measurePath);
            },

            set: function(measureNameOrMeasuresObj, measure) {
                dot.set(this._log, measureNameOrMeasuresObj, measure, 'factorized');
                return this;
            },

            _applyProcessor: function(processor) {
                return processor.apply(this, [this].concat(this.processorParams));
            },

            _applyProcessors: function() {
                var nextProcToApply = this.processors,
                    procToApply,
                    applied,
                    process,
                    procNum,
                    result;
                do {
                    applied = 0;
                    procToApply = nextProcToApply;
                    nextProcToApply = [];
                    for (procNum in procToApply) {
                        process = procToApply[procNum];
                        // apply the precessor
                        result = this._applyProcessor(process);
                        if (tools.isUnset(result) || result) {
                            applied++;
                        } else {
                            nextProcToApply.push(process);
                        }
                    }
                } while (applied > 0 && nextProcToApply.length > 0);
            },

            export: function() {
                var log = {};
                this._applyProcessors();
                $.extend(true, log, this._log);
                return log;
            },

            getToken: function() {
                this._givenToken++;
            },

            releaseToken: function() {
                this._givenToken--;
            },

            isFree: function() {
                return this._givenToken <= 0;
            },

        };
    });


});
