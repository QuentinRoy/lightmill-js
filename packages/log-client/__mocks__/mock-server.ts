import {
  http,
  HttpHandler,
  HttpResponse,
  HttpResponseResolver,
  RequestHandler,
} from 'msw';
import { setupServer, SetupServerApi } from 'msw/node';
import { RequiredKeysOf } from 'type-fest';
import { test } from 'vitest';
import { paths } from '../generated/api.js';

type ApiReponseCode<
  Path extends keyof paths,
  Method extends Exclude<RequiredKeysOf<paths[Path]>, 'parameters'>,
> = paths extends {
  [P in Path]: { [M in Method]: { responses: infer Responses } };
}
  ? keyof Responses
  : never;

type ApiResponse<
  Path extends keyof paths,
  Method extends Exclude<RequiredKeysOf<paths[Path]>, 'parameters'>,
  Status extends ApiReponseCode<Path, Method> = ApiReponseCode<Path, Method>,
> = paths extends {
  [P in Path]: {
    [M in Method]: {
      responses: {
        [S in Status]: { content: { 'application/json': infer R } };
      };
    };
  };
}
  ? R
  : never;

type ApiRequestBody<
  Path extends keyof paths,
  Method extends Exclude<RequiredKeysOf<paths[Path]>, 'parameters'>,
> = paths extends {
  [P in Path]: {
    [M in Method]: {
      requestBody: { content: { 'application/json': infer R } };
    };
  };
}
  ? R
  : never;

export class MockServer {
  #baseUrl = 'https://server.test/api';
  #runs = new Map<string, Run>();
  #experiments = new Map<string, Experiment>();
  #server: SetupServerApi;
  #requests: Array<Request> = [];

  getBaseUrl(): string {
    return this.#baseUrl;
  }

  constructor() {
    this.#server = setupServer(...this.#getHandlers());
    this.#server.events.on('request:start', ({ request }) => {
      // We need to clone the request because it is a stream and may be
      // consumed by the server.
      this.#requests.push(request.clone());
    });
  }

  #getHandlers(): RequestHandler[] {
    const { get, delete: del, post, patch } = createMethods(this.#baseUrl);
    return [
      get('/experiments/:experimentId', ({ params }) => {
        const experiment = this.#experiments.get(params.experimentId as string);
        if (experiment == null) {
          return HttpResponse.json(
            { errors: [] } satisfies ApiResponse<
              '/experiments/{id}',
              'get',
              404
            >,
            { status: 404 },
          );
        }
        return HttpResponse.json(
          {
            data: {
              id: experiment.id,
              type: 'experiments',
              attributes: { name: experiment.name },
            },
          } satisfies ApiResponse<'/experiments/{id}', 'get', 200>,
          { status: 200 },
        );
      }),

      get('/experiments', ({ request }) => {
        const queryParams = parseUrlQuery(request.url);
        return HttpResponse.json(
          {
            data: Array.from(this.#experiments.values())
              .filter((e) => {
                let nameFilter = queryParams['filter[name]'];
                if (nameFilter == null) {
                  return true;
                }
                return nameFilter.includes(e.name);
              })
              .map((experiment) => ({
                id: experiment.id,
                type: 'experiments',
                attributes: { name: experiment.name },
              })),
          } satisfies ApiResponse<'/experiments', 'get', 200>,
          { status: 200 },
        );
      }),

      get('/sessions/current', () =>
        HttpResponse.json({
          data: {
            type: 'sessions',
            id: 'current',
            attributes: { role: 'participant' },
            relationships: {
              runs: {
                data: Array.from(this.#runs.values(), (r) => ({
                  type: 'runs',
                  id: r.runId,
                })),
              },
            },
          },
        } satisfies ApiResponse<'/sessions/{id}', 'get', 200>),
      ),

      del('/sessions/current', () =>
        HttpResponse.json({ data: null } satisfies ApiResponse<
          '/sessions/{id}',
          'delete',
          200
        >),
      ),

      get('/runs', async ({ request }) => {
        const queryParams = parseUrlQuery(request.url);
        return HttpResponse.json(
          {
            data: Array.from(this.#runs.values())
              .filter((r) => {
                let statusFilter = queryParams['filter[status]'];
                if (statusFilter == null) {
                  return true;
                }
                return statusFilter.includes(r.runStatus);
              })
              .map((r) => ({
                type: 'runs',
                id: r.runId,
                attributes: {
                  status: r.runStatus,
                  name: r.runName,
                  lastLogNumber: Math.max(
                    0,
                    ...r.lastLogs.map((log) => log.number),
                  ),
                },
                relationships: {
                  experiment: {
                    data: { type: 'experiments', id: r.experimentId },
                  },
                  lastLogs: {
                    data: r.lastLogs.map((log) => ({
                      type: 'logs',
                      id: log.id,
                    })),
                  },
                },
              })),
            included: [
              ...Array.from(this.#runs.values()).flatMap((r) =>
                r.lastLogs.map((log) => ({
                  type: 'logs' as const,
                  id: log.id,
                  attributes: {
                    logType: log.type,
                    number: log.number,
                    values: log.values,
                  },
                  relationships: {
                    run: { data: { type: 'runs' as const, id: r.runId } },
                  },
                })),
              ),
              ...Array.from(this.#experiments.values()).map((e) => {
                return {
                  type: 'experiments' as const,
                  id: e.id,
                  attributes: { name: e.name },
                };
              }),
            ],
          } satisfies ApiResponse<'/runs', 'get', 200>,
          { status: 200 },
        );
      }),

      post('/runs', async () => {
        return HttpResponse.json(
          {
            data: { type: 'runs', id: 'default-run-id' },
          } satisfies ApiResponse<'/runs', 'post', 201>,
          { status: 201 },
        );
      }),

      get('/runs/:runId', ({ params, request }) => {
        const run = this.#runs.get(params.runId as string);
        if (run == null) {
          return HttpResponse.json(
            { errors: [] } satisfies ApiResponse<'/runs/{id}', 'get', 404>,
            { status: 404 },
          );
        }
        let included: ApiResponse<'/runs/{id}', 'get', 200>['included'];
        let urlParams = parseUrlQuery(request.url);
        if (urlParams.include?.includes('experiment')) {
          included = included ?? [];
          included.push({
            type: 'experiments',
            id: run.experimentId,
            attributes: {
              name:
                this.#experiments.get(run.experimentId)?.name ??
                'Unknown Experiment',
            },
          });
        }
        if (urlParams.include?.includes('lastLogs')) {
          included = included ?? [];
          included.push(
            ...run.lastLogs.map((log) => ({
              type: 'logs' as const,
              id: log.id,
              attributes: {
                logType: log.type,
                number: log.number,
                values: log.values,
              },
              relationships: {
                run: { data: { type: 'runs' as const, id: run.runId } },
              },
            })),
          );
        }
        return HttpResponse.json({
          data: {
            type: 'runs',
            id: run.runId,
            attributes: {
              name: run.runName,
              status: run.runStatus,
              lastLogNumber: Math.max(...run.lastLogs.map((l) => l.number)),
            },
            relationships: {
              experiment: {
                data: { type: 'experiments', id: run.experimentId },
              },
              lastLogs: {
                data: run.lastLogs.map((log) => ({ type: 'logs', id: log.id })),
              },
            },
          },
          included,
        } satisfies ApiResponse<'/runs/{id}', 'get', 200>);
      }),

      patch('/runs/:id', async ({ params, request }) => {
        const body = (await request.json()) as ApiRequestBody<
          '/runs/{id}',
          'patch'
        >;
        const run = this.#runs.get(params.id as string);
        if (run == null) {
          return HttpResponse.json(
            { errors: [] } satisfies ApiResponse<'/runs/{id}', 'patch', 404>,
            { status: 404 },
          );
        }
        run.runStatus = body.data.attributes?.status ?? run.runStatus;
        const lastLogNumber = body.data.attributes?.lastLogNumber;
        if (lastLogNumber != null) {
          run.lastLogs = run.lastLogs.filter((l) => l.number <= lastLogNumber);
        }

        return HttpResponse.json({
          data: {
            id: run.runId,
            type: 'runs' as const,
            attributes: {
              status: run.runStatus,
              lastLogNumber: Math.max(0, ...run.lastLogs.map((l) => l.number)),
            },
            relationships: {
              experiment: {
                data: { id: run.experimentId, type: 'experiments' },
              },
              lastLogs: {
                data: run.lastLogs.map((l) => ({ id: l.id, type: 'logs' })),
              },
            },
          },
        } satisfies ApiResponse<'/runs/{id}', 'patch', 200>);
      }),

      post('/logs', () => {
        return HttpResponse.json({
          data: { id: 'log-id', type: 'logs' },
        } satisfies ApiResponse<'/logs', 'post', 201>);
      }),
    ];
  }

  async waitForRequests() {
    return Promise.all(
      this.#requests.map(async (r) => {
        let body = r.body == null ? null : await r.json();
        return { method: r.method, url: r.url, body };
      }),
    );
  }

  // In most cases, these are the ones we care about because these are the ones
  // that actually change the state of the server.
  static #HTTP_CHANGE_METHODS = ['PUT', 'PATCH', 'POST', 'DELETE'];
  async waitForChangeRequests() {
    let requests = await this.waitForRequests();
    return requests.filter((r) =>
      MockServer.#HTTP_CHANGE_METHODS.includes(r.method),
    );
  }

  set(
    entries: Array<
      | (Omit<Run, 'lastLogs'> & {
          experimentName?: string;
          lastLogs?: Array<{
            type: string;
            number: number;
            id?: string;
            values?: Record<string, unknown> & { date?: Date };
          }>;
        })
      | { experimentName?: string; experimentId: string }
    >,
  ) {
    this.#runs = new Map();
    this.#experiments = new Map();
    for (let i = 0; i < entries.length; i++) {
      let entry = entries[i];
      let experimentName: string =
        entry.experimentName != null
          ? entry.experimentName
          : `${entry.experimentId}-name`;
      if ('runId' in entry) {
        let { lastLogs, ...run } = entry;
        let runName = run.runName ?? `${run.runId}-name`;
        this.#runs.set(run.runId, {
          ...run,
          runName,
          lastLogs:
            lastLogs?.map((l) => {
              let id = l.id ?? `log-${run.runId}-${l.number}`;
              let date = l.values?.date ?? new Date();
              let values = { ...l.values, date: date.toISOString() };
              return { ...l, id, values };
            }) ?? [],
        });
        if (!this.#experiments.has(entry.experimentId)) {
          this.#experiments.set(entry.experimentId, {
            id: entry.experimentId,
            name: experimentName,
          });
        }
        continue;
      }
      this.#experiments.set(entry.experimentId, {
        id: entry.experimentId,
        name: experimentName,
      });
    }
  }

  reset() {
    this.#requests = [];
    this.#runs = new Map();
    this.#experiments = new Map();
  }

  start() {
    this.#server.listen({ onUnhandledRequest: 'error' });
  }

  stop() {
    this.#server.close();
  }
}

function createMethods(baseUrl: string) {
  let result: Record<
    string,
    (p: string, handler: HttpResponseResolver) => HttpHandler
  > = {};
  for (let method of ['get', 'put', 'post', 'patch', 'delete'] as const) {
    result[method] = (path, handler) =>
      http[method](`${baseUrl}${path}`, handler);
  }

  return result;
}

type Run = {
  runId: string;
  runName?: string;
  experimentId: string;
  runStatus: 'idle' | 'completed' | 'canceled' | 'running' | 'interrupted';
  lastLogs: Array<{
    id: string;
    type: string;
    number: number;
    values: Record<string, unknown> & { date: string };
  }>;
};

type Experiment = { id: string; name: string };

interface Fixture {
  server: MockServer;
}

export const serverTest = test.extend<Fixture>({
  // eslint-disable-next-line no-empty-pattern
  server: async ({}, use) => {
    let server = new MockServer();
    server.start();
    await use(server);
    server.stop();
  },
});

function parseUrlQuery(url: string) {
  let query = url.split('?')[1];
  if (query == null) return {};
  return query.split('&').reduce<Record<string, string[]>>((acc, pair) => {
    let [key, value] = pair.split('=');
    let values = acc[key] || [];
    values.push(value);
    acc[key] = values;
    return acc;
  }, {});
}
