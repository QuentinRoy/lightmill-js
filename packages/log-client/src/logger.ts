import type { paths } from '@lightmill/log-api';
import type { Client as FetchClient } from 'openapi-fetch';
import type { JsonValue } from 'type-fest';
import type { LogValuesSerializer, RunStatus } from './types.js';
import { apiMediaType, RequestError } from './utils.js';

interface Typed<Type extends string = string> {
  type: Type;
}
interface OptionallyDated {
  date?: Date;
}
interface JsonObjectAndDate {
  [key: string]: JsonValue | Date | undefined | JsonObjectAndDate;
}
interface AnyLog extends Typed, OptionallyDated, JsonObjectAndDate {}

export class LightmillLogger<
  ClientLog extends Typed & OptionallyDated = AnyLog,
> {
  #serializeValues: LogValuesSerializer<ClientLog>;
  #runId: string;
  #runStatus: RunStatus = 'running';
  #lastLogNumber: number;
  #fetchClient: FetchClient<paths, `${string}/${string}`>;
  #pendingLogs = new Set<number>();
  #emptyQueueCallback: null | (() => void) = null;
  #error: Error | null = null;

  constructor({
    runId,
    serializeLog,
    lastLogNumber,
    fetchClient,
  }: {
    runId: string;
    lastLogNumber: number;
    fetchClient: FetchClient<paths, `${string}/${string}`>;
    serializeLog: LogValuesSerializer<ClientLog>;
  }) {
    this.#serializeValues = serializeLog;
    this.#runId = runId;
    this.#fetchClient = fetchClient;
    this.#lastLogNumber = lastLogNumber;
  }

  async addLog({ type, ...values }: ClientLog) {
    if (this.#runStatus !== 'running') {
      throw new Error(
        `Can only add logs when logger is running. Logger is ${this.#runStatus}`,
      );
    }
    if (type == null) {
      throw new Error(
        'Trying to add a log without a type. Logs must have a type',
      );
    }
    const logNumber = this.#lastLogNumber + 1;
    this.#lastLogNumber = logNumber;
    this.#addPendingLog(logNumber);
    let error: Error | null = null;
    try {
      let response = await this.#fetchClient.POST('/logs', {
        credentials: 'include',
        headers: { 'content-type': apiMediaType },
        body: {
          data: {
            type: 'logs',
            attributes: {
              logType: type,
              number: logNumber,
              values: this.#serializeValues({ date: new Date(), ...values }),
            },
            relationships: { run: { data: { type: 'runs', id: this.#runId } } },
          },
        },
      });
      error = response.error == null ? null : new RequestError(response);
    } catch (caughtError) {
      error =
        caughtError instanceof Error
          ? caughtError
          : new Error(String(caughtError));
    }
    this.#error = this.#error == null ? error : this.#error;
    this.#removePendingLog(logNumber);
    if (this.#pendingLogs.size === 0) this.#emptyQueueCallback?.();
    if (error != null) throw error;
  }

  async #addPendingLog(logNumber: number) {
    this.#pendingLogs.add(logNumber);
  }

  async #removePendingLog(logNumber: number) {
    this.#pendingLogs.delete(logNumber);
  }

  #flushPromise: Promise<void> | null = null;
  async flush() {
    if (this.#pendingLogs.size === 0) return;
    if (!this.#flushPromise) {
      this.#flushPromise = new Promise<void>((resolve) => {
        this.#emptyQueueCallback = () => {
          this.#flushPromise = null;
          this.#emptyQueueCallback = null;
          resolve();
        };
      });
    }
    await this.#flushPromise;
    if (this.#error) {
      throw this.#error;
    }
  }

  async completeRun() {
    await this.#endRun('completed');
  }

  async cancelRun() {
    await this.#endRun('canceled');
  }

  async interruptRun() {
    await this.#endRun('interrupted');
  }

  async #endRun(runStatus: 'canceled' | 'completed' | 'interrupted') {
    if (this.#runStatus !== 'running') {
      throw new Error(
        `Cannot end a run that is not running. Run is ${runStatus}`,
      );
    }
    await this.flush();
    let response = await this.#fetchClient.PATCH('/runs/{id}', {
      credentials: 'include',
      params: { path: { id: this.#runId } },
      headers: { 'content-type': apiMediaType },
      body: {
        data: {
          type: 'runs',
          id: this.#runId,
          attributes: { status: runStatus },
        },
      },
    });
    if (response.error) {
      throw new RequestError(response);
    }
    this.#runStatus = runStatus;
  }
}

export type Logger = LightmillLogger;
