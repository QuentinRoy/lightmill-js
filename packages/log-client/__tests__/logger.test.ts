import type { paths } from '@lightmill/log-api';
import createClient from 'openapi-fetch';
import { describe, expect, vi } from 'vitest';
import { serverTest } from '../__mocks__/mock-server.js';
import { LightmillLogger } from '../src/logger.js';

const it = serverTest.extend<{
  timer: void;
  logger: LightmillLogger;
  run: {
    experimentName: string;
    runName: string;
    experimentId: string;
    runId: string;
  };
  resumedLogger: LightmillLogger;
  resumedLogCount: number;
}>({
  timer: [
    // eslint-disable-next-line no-empty-pattern
    async ({}, use) => {
      vi.useFakeTimers({
        toFake: [
          'Date',
          'setTimeout',
          'clearTimeout',
          'setInterval',
          'clearInterval',
        ],
      });
      await use();
      vi.useRealTimers();
    },
    { auto: true },
  ],
  run: Object.freeze({
    experimentName: 'exp-name',
    experimentId: 'exp-id',
    runName: 'run-name',
    runId: 'run-id',
  }),
  logger: async ({ server, run }, use) => {
    const fetchClient = createClient<paths>({
      baseUrl: server.getBaseUrl(),
      headers: { accept: 'application/json' },
    });
    const logger = new LightmillLogger({
      fetchClient,
      ...run,
      lastLogNumber: 0,
      serializeLog: (x) => JSON.parse(JSON.stringify(x)),
    });
    server.set([run]);
    await use(logger);
    server.reset();
  },
  resumedLogCount: 100,
  resumedLogger: async ({ server, resumedLogCount, run }, use) => {
    const fetchClient = createClient<paths>({ baseUrl: server.getBaseUrl() });
    const logger = new LightmillLogger({
      fetchClient,
      ...run,
      lastLogNumber: resumedLogCount,
      serializeLog: (x) => JSON.parse(JSON.stringify(x)),
    });
    server.set([run]);
    await use(logger);
    server.reset();
  },
});

describe('LogClient#addLog', () => {
  it('should send one log', async ({ logger, server, expect }) => {
    await logger.addLog({
      type: 'mock-log',
      val: 1,
      date: new Date('2021-06-03T02:00:00.000Z'),
    });
    await expect(server.waitForChangeRequests()).resolves
      .toMatchInlineSnapshot(`
      [
        {
          "body": {
            "data": {
              "attributes": {
                "logType": "mock-log",
                "number": 1,
                "values": {
                  "date": "2021-06-03T02:00:00.000Z",
                  "val": 1,
                },
              },
              "relationships": {
                "run": {
                  "data": {
                    "id": "run-id",
                    "type": "runs",
                  },
                },
              },
              "type": "logs",
            },
          },
          "method": "POST",
          "url": "https://server.test/api/logs",
        },
      ]
    `);
  });

  it('should add a default date to logs ', async ({ logger, server }) => {
    vi.useFakeTimers({ toFake: ['Date'] });
    vi.setSystemTime('2019-06-03T02:00:00.000Z');
    await logger.addLog({ type: 'mock-log', val: 'xxx' });
    await expect(server.waitForChangeRequests()).resolves
      .toMatchInlineSnapshot(`
        [
          {
            "body": {
              "data": {
                "attributes": {
                  "logType": "mock-log",
                  "number": 1,
                  "values": {
                    "date": "2019-06-03T02:00:00.000Z",
                    "val": "xxx",
                  },
                },
                "relationships": {
                  "run": {
                    "data": {
                      "id": "run-id",
                      "type": "runs",
                    },
                  },
                },
                "type": "logs",
              },
            },
            "method": "POST",
            "url": "https://server.test/api/logs",
          },
        ]
    `);
  });

  it('should send logs with no provided values', async ({ logger, server }) => {
    vi.useFakeTimers({ toFake: ['Date'] });
    vi.setSystemTime('2019-06-03T02:00:00.000Z');
    await logger.addLog({ type: 'mock-log' });
    await expect(server.waitForChangeRequests()).resolves
      .toMatchInlineSnapshot(`
      [
        {
          "body": {
            "data": {
              "attributes": {
                "logType": "mock-log",
                "number": 1,
                "values": {
                  "date": "2019-06-03T02:00:00.000Z",
                },
              },
              "relationships": {
                "run": {
                  "data": {
                    "id": "run-id",
                    "type": "runs",
                  },
                },
              },
              "type": "logs",
            },
          },
          "method": "POST",
          "url": "https://server.test/api/logs",
        },
      ]
    `);
  });
});

describe('LogClient#addLog (after resume)', () => {
  it('should properly start numbering after a run has been resumed', async ({
    server,
  }) => {
    server.set([
      {
        experimentId: 'test-experiment',
        runId: 'test-run',
        runStatus: 'running',
        lastLogs: [
          { type: 'test-type', number: 4 },
          { type: 'other-type', number: 6 },
        ],
      },
    ]);
    const logger = new LightmillLogger({
      fetchClient: createClient({
        baseUrl: 'https://server.test/api',
        headers: { contentType: 'application/vnd.api+json' },
      }),
      runId: 'test-run',
      lastLogNumber: 4,
      serializeLog: (log) => JSON.parse(JSON.stringify(log)),
    });
    await Promise.all([
      logger.addLog({
        type: 'mock-log',
        val: 'a',
        date: new Date('2021-06-03T02:00:00.000Z'),
      }),
      logger.addLog({
        type: 'mock-log',
        val: 'b',
        date: new Date('2021-06-03T02:00:10.000Z'),
      }),
      logger.addLog({
        type: 'mock-log',
        val: 'c',
        date: new Date('2021-06-03T02:00:20.000Z'),
      }),
    ]);
    await expect(server.waitForChangeRequests()).resolves.toMatchSnapshot();
  });
});

describe('LogClient#flush', () => {
  it('should flush', async ({ logger, server }) => {
    logger.addLog({
      type: 'mock-log',
      val: 1,
      date: new Date('2021-06-03T02:00:00.000Z'),
    });
    logger.addLog({
      type: 'mock-log',
      val: 2,
      date: new Date('2021-06-03T03:00:00.000Z'),
    });
    logger.addLog({
      type: 'mock-log',
      val: 3,
      date: new Date('2021-06-03T04:00:00.000Z'),
    });
    await logger.flush();
    await expect(server.waitForChangeRequests()).resolves.toMatchSnapshot();
  });

  it('should flush even if flush is called multiple times', async ({
    logger,
    server,
  }) => {
    logger.addLog({
      type: 'mock-log',
      val: 1,
      date: new Date('2021-06-03T02:00:00.000Z'),
    });
    logger.addLog({
      type: 'mock-log',
      val: 2,
      date: new Date('2021-06-03T03:00:00.000Z'),
    });
    logger.addLog({
      type: 'mock-log',
      val: 3,
      date: new Date('2021-06-03T04:00:00.000Z'),
    });
    let flush1 = logger.flush();
    let flush2 = logger.flush();
    let flush3 = logger.flush();
    await expect(
      Promise.all([flush1, flush2, flush3]),
    ).resolves.toMatchSnapshot();
    await expect(server.waitForChangeRequests()).resolves.toMatchSnapshot();
  });
});

describe('LogClient#completeRun', () => {
  it('should complete', async ({ server, logger }) => {
    await logger.completeRun();
    await expect(server.waitForChangeRequests()).resolves
      .toMatchInlineSnapshot(`
      [
        {
          "body": {
            "data": {
              "attributes": {
                "status": "completed",
              },
              "id": "run-id",
              "type": "runs",
            },
          },
          "method": "PATCH",
          "url": "https://server.test/api/runs/run-id",
        },
      ]
    `);
  });
});

describe('LogClient#cancelRun', () => {
  it('should cancel', async ({ server, logger }) => {
    await logger.cancelRun();
    await expect(server.waitForChangeRequests()).resolves
      .toMatchInlineSnapshot(`
      [
        {
          "body": {
            "data": {
              "attributes": {
                "status": "canceled",
              },
              "id": "run-id",
              "type": "runs",
            },
          },
          "method": "PATCH",
          "url": "https://server.test/api/runs/run-id",
        },
      ]
    `);
  });
});
