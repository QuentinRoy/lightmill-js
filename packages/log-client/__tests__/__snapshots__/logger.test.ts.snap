// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogClient#addLog (after resume) > should properly start numbering after a run has been resumed 1`] = `
[
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 5,
          "values": {
            "date": "2021-06-03T02:00:00.000Z",
            "val": "a",
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "test-run",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 6,
          "values": {
            "date": "2021-06-03T02:00:10.000Z",
            "val": "b",
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "test-run",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 7,
          "values": {
            "date": "2021-06-03T02:00:20.000Z",
            "val": "c",
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "test-run",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
]
`;

exports[`LogClient#flush > should flush 1`] = `
[
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 1,
          "values": {
            "date": "2021-06-03T02:00:00.000Z",
            "val": 1,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 2,
          "values": {
            "date": "2021-06-03T03:00:00.000Z",
            "val": 2,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 3,
          "values": {
            "date": "2021-06-03T04:00:00.000Z",
            "val": 3,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
]
`;

exports[`LogClient#flush > should flush even if flush is called multiple times 1`] = `
[
  undefined,
  undefined,
  undefined,
]
`;

exports[`LogClient#flush > should flush even if flush is called multiple times 2`] = `
[
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 1,
          "values": {
            "date": "2021-06-03T02:00:00.000Z",
            "val": 1,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 2,
          "values": {
            "date": "2021-06-03T03:00:00.000Z",
            "val": 2,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
  {
    "body": {
      "data": {
        "attributes": {
          "logType": "mock-log",
          "number": 3,
          "values": {
            "date": "2021-06-03T04:00:00.000Z",
            "val": 3,
          },
        },
        "relationships": {
          "run": {
            "data": {
              "id": "run-id",
              "type": "runs",
            },
          },
        },
        "type": "logs",
      },
    },
    "method": "POST",
    "url": "https://server.test/api/logs",
  },
]
`;
